### PR TITLE
Jfb 4

### PR DESCRIPTION
**Pull request by Sialenius**

### CODE REVIEW

File: src/Model/Enums/ConcertHall.java [hunk 1]
```diff
@@ -1,4 +1,4 @@
-package Model;
+package Model.Enums;
 
 public enum ConcertHall {
```
Comment (AtticSiol) [inline src/Model/Enums/ConcertHall.java]: in upper case

File: src/Model/Ticket.java [hunk 2]
```diff
@@ -97,88 +68,88 @@ public void setStadiumSector(StadiumSector stadiumSector) {
...
         else {
-            System.out.println("Price: -");
+            return false;
         }
-        System.out.println("----------------");
+    }
 
+  @Override
+    public int hashCode() {
+        return getId().hashCode();
```
Comment (AtticSiol) [inline src/Model/Ticket.java]: check it

File: src/Model/Ticket.java [hunk 3]
```diff
@@ -1,81 +1,55 @@
...
 import java.time.LocalDateTime;
-import java.time.chrono.ChronoLocalDateTime;
 import java.time.format.DateTimeFormatter;
-import java.time.format.DateTimeFormatterBuilder;
-import java.time.format.FormatStyle;
 
-public class Ticket {
-    private static int ticketsCounter = 1;
-    private final LocalDateTime ticketCreationTime = LocalDateTime.now();
+public class Ticket implements ID, Printer {
```
Comment (AtticSiol) [inline src/Model/Ticket.java]: follow the correct order: constants, variables, constructors, setter/getters, equals+hashCode, toString

File: src/Model/Enums/ConcertHall.java (hunk 1 above)
Comment (Sialenius) [reply src/Model/Enums/ConcertHall.java]: Fixed. I rewrote variables of this enum in upper case + added the field name, constructor and getter to represent the enum values in different way

File: src/Model/Ticket.java (hunk 3 above)
Comment (Sialenius) [reply src/Model/Ticket.java]: Fixed. I refactored the class in proper order: constants, variables, constructor, etc.

File: src/Control/TicketService.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
+package Control;
+
+import Model.*;
```
Comment (AtticSiol) [inline src/Control/TicketService.java]: use full separate name of imported classes, go not group them with "*"

File: src/Model/User.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package Model;
+
+import View.Printable;
+
+public abstract class User implements ID, Printable {
```
Comment (AtticSiol) [inline src/Model/User.java]: can't find interface ID into your code

File: src/Model/Ticket.java [hunk 6]
```diff
@@ -1,81 +1,73 @@
...
 import java.time.LocalDateTime;
-import java.time.chrono.ChronoLocalDateTime;
 import java.time.format.DateTimeFormatter;
-import java.time.format.DateTimeFormatterBuilder;
-import java.time.format.FormatStyle;
 
-public class Ticket {
-    private static int ticketsCounter = 1;
-    private final LocalDateTime ticketCreationTime = LocalDateTime.now();
+public class Ticket implements ID, Printable {
```
Comment (AtticSiol) [inline src/Model/Ticket.java]: the same, can't find interface ID. Perhaps you replaced it by Entity abstract class

File: src/Model/Ticket.java [hunk 7]
```diff
@@ -1,81 +1,73 @@
...
+    //@Min(value = 0, message = "Invalid backpack weight")
     private double backpackAllowedWeight;
+
+   // @Min(value = 0, message = "Invalid price")
     private BigDecimal price;
 
-    public int getTicketId() {
-        return ticketId;
+    public Ticket() {
+        this(ConcertHall.NOT_SPECIFIED, 3, null, false, StadiumSector.NOT_SPECIFIED, 0, 0); // How to create a Ticket without Event Time, not using "null"??
```
Comment (AtticSiol) [inline src/Model/Ticket.java]: you can implement here any logic:
1. start date of Timestamp is 1970-01-01 01:00:00.0
so, you can use it as default "unspecified" value (new Timestamp(0))
2. or you can define that default value will be + 5 days:
Timestamp.valueOf(LocalDateTime.now().plusDays(5))

File: src/Model/Ticket.java [hunk 8]
```diff
@@ -97,88 +86,69 @@ public void setStadiumSector(StadiumSector stadiumSector) {
...
-        this(concertHall, eventCode, eventTime, false, StadiumSector.not_specified, 0);
+    public void share(PhoneNumber phoneNumber) {
+        System.out.println("Ticket "  + getId() + " was shared by phone number");
     }
 
-    public Ticket(ConcertHall concertHall, int eventCode, LocalDateTime eventTime, boolean isPromo, StadiumSector stadiumSector, double backpackAllowedWeight) {
-        ticketId = ticketsCounter;
-        ticketsCounter++;
+    public void share(PhoneNumber phoneNumber, Email email) {
+        System.out.println("Ticket " + getId() + " was shared by phone number and email");
```
Comment (AtticSiol) [inline src/Model/Ticket.java]: if you declare arguments, you need to do smth with them inside the method. For example, print them also in your message.

File: src/Model/Ticket.java (hunk 2 above)
Comment (Sialenius) [reply src/Model/Ticket.java]: Fixed. I updated the hashCode() method

File: src/Control/TicketService.java (hunk 4 above)
Comment (Sialenius) [reply src/Control/TicketService.java]: Fixed. I updated import and added classes from package "Model" separately

File: src/Model/Ticket.java (hunk 8 above)
Comment (Sialenius) [reply src/Model/Ticket.java]: Fixed. I updated overloaded methods share() to use all input parameters

File: src/Model/Ticket.java (hunk 7 above)
Comment (Sialenius) [reply src/Model/Ticket.java]: Fixed. Thank you for advice!

File: src/Model/User.java (hunk 5 above)
Comment (Sialenius) [reply src/Model/User.java]: Fixed. I replaced using ID interface to extending abstract class Entity

File: src/Model/Ticket.java (hunk 6 above)
Comment (Sialenius) [reply src/Model/Ticket.java]: Fixed. I replaced using ID interface to extending abstract class Entity

File: src/Model/Ticket.java [hunk 9]
```diff
@@ -1,81 +1,74 @@
...
-
-    private static final SimpleDateFormat TIME_FORMAT = new SimpleDateFormat("HH:mm");
-    private static final DecimalFormat FORMATTER = new DecimalFormat("$#,##0.00");
-
-    private int ticketId = 0;
-    private ConcertHall concertHall;
-    private int eventCode;
+
+
+public class Ticket extends Entity implements Printable, AnnaotationAnalizator {
```
Comment (AtticSiol) [inline src/Model/Ticket.java]: AnnaotationAnalizator - suggest it is not necessary

File: src/Model/Ticket.java (hunk 7 above)
Comment (AtticSiol) [reply src/Model/Ticket.java]: can't find the solution into code

File: src/Model/Enums_and_constants/AppConstants.java [hunk 10]
```diff
@@ -0,0 +1,15 @@
+package Model.Enums_and_constants;
```
Comment (AtticSiol) [inline src/Model/Enums_and_constants/AppConstants.java]: name all packages into lower case: model, control etc.
do not use in packages names "_"
store enums into "enums" package, constants - into "constant" package or into the package, where these constants are used (for example directly into model package)

**Review by AtticSiol — APPROVED** [review]

**Comment by AtticSiol** [comment]
approved. before merging remove all redundant comments

**Pull request merged**